pkg/shp/follower/follower.go

```diff
@@ -0,0 +1,219 @@
+package follower
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/shipwright-io/cli/pkg/shp/params"
+	"github.com/shipwright-io/cli/pkg/shp/reactor"
+	"github.com/shipwright-io/cli/pkg/shp/tail"
+	"github.com/shipwright-io/cli/pkg/shp/util"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/client-go/kubernetes"
+)
+
+// PodLogsFollower uses PodWatcher to react upon pod state changes with the final objective of
+// streaming all logs from all the containers in the pod.
+type PodLogsFollower struct {
+	ctx context.Context
+
+	pw        *reactor.PodWatcher          // podwatcher instance
+	clientset kubernetes.Interface         // kubernetes client
+	ioStreams *genericclioptions.IOStreams // log output interface
+
+	podIsRunning   bool            // pod is set as running
+	logTail        *tail.Tail      // helper to tail container logs
+	logTailStarted map[string]bool // containers with logTail started
+
+	onlyOnce bool // retrieve the pod logs only once
+}
+
+var (
+	// ErrPodDeleted a pod has been deleted, having DeletionTimestamp set
+	ErrPodDeleted = errors.New("pod has been deleted")
+
+	// ErrPodFailed a pod has failed, generic reason
+	ErrPodFailed = errors.New("pod has failed")
+
+	// ErrPodStatusUnknown a pod has status condition set to unknown
+	ErrPodStatusUnknown = errors.New("pod status is unknown")
+)
+
+// log log entries on the standard output of the ioStreams instance.
+func (p *PodLogsFollower) log(format string, a ...interface{}) {
+	fmt.Fprintf(p.ioStreams.Out, format, a...)
+}
+
+// log log entries on the error output of the ioStreams instance.
+func (p *PodLogsFollower) logError(format string, a ...interface{}) {
+	fmt.Fprintf(p.ioStreams.ErrOut, format, a...)
+}
+
+// onPodRunning as soon as the pod gets into the running state it will start streaming logs from all
+// containers, and mark it as running.
+func (p *PodLogsFollower) onPodRunning(pod *corev1.Pod) {
+	if p.onlyOnce {
+		p.getPodLogs(pod)
+		return
+	}
+	// when the pod is already running the log tailing must be already set up, thefore skipping it
+	// when dealing with subsequent pod modifications
+	if p.podIsRunning {
+		return
+	}
+
+	p.log("Pod %q in %q state, starting up log tail\n", pod.GetName(), corev1.PodRunning)
+
+	time.Sleep(3 * time.Second)
```

Comment:
I assume we wait here so that the user can read the message that was just printed? Is worth a comment, I think.